source/opt/combine_access_chains.cpp

```diff
@@ -0,0 +1,253 @@
+    Instruction* index_inst =
+        def_use_mgr->GetDef(inst->GetSingleWordInOperand(i));
+    const analysis::Constant* index_constant =
+        context()->get_constant_mgr()->GetConstantFromInst(index_inst);
+    if (index_constant) {
+      uint32_t index_value = GetConstantValue(index_constant);
+      element_indices.push_back(index_value);
+    } else {
+      // This index must not matter to resolve the type in valid SPIR-V.
+      element_indices.push_back(0);
+    }
+  }
+  type = type_mgr->GetMemberType(type, element_indices);
+  return type;
+}
+
+bool CombineAccessChains::CombineIndices(Instruction* ptr_input,
+                                         Instruction* inst,
+                                         std::vector<Operand>* new_operands) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  analysis::ConstantManager* constant_mgr = context()->get_constant_mgr();
+
+  Instruction* last_index_inst = def_use_mgr->GetDef(
+      ptr_input->GetSingleWordInOperand(ptr_input->NumInOperands() - 1));
+  const analysis::Constant* last_index_constant =
+      constant_mgr->GetConstantFromInst(last_index_inst);
+
+  Instruction* element_inst =
+      def_use_mgr->GetDef(inst->GetSingleWordInOperand(1));
+  const analysis::Constant* element_constant =
+      constant_mgr->GetConstantFromInst(element_inst);
+
+  // Combine the last index of the AccessChain (|ptr_inst|) with the element
+  // operand of the PtrAccessChain (|inst|).
+  const bool combining_element_operands =
+      inst->opcode() == SpvOpPtrAccessChain &&
+      ptr_input->opcode() == SpvOpPtrAccessChain &&
+      ptr_input->NumInOperands() == 2;
+  uint32_t new_value_id = 0;
+  const analysis::Type* type = GetIndexedType(ptr_input);
+  if (last_index_constant && element_constant) {
+    // Combine the constants.
+    uint32_t new_value = GetConstantValue(last_index_constant) +
+                         GetConstantValue(element_constant);
+    const analysis::Constant* new_value_constant =
+        constant_mgr->GetConstant(last_index_constant->type(), {new_value});
+    Instruction* new_value_inst =
+        constant_mgr->GetDefiningInstruction(new_value_constant);
+    new_value_id = new_value_inst->result_id();
+  } else if (!type->AsStruct() || combining_element_operands) {
+    // TODO(alan-baker): handle this unlikely case.
+    if (last_index_inst->type_id() != element_inst->type_id()) return false;
+
+    // Generate an addition of the two indices.
+    InstructionBuilder builder(
+        context(), inst,
+        IRContext::kAnalysisDefUse | IRContext::kAnalysisInstrToBlockMapping);
+    Instruction* addition = builder.AddIAdd(last_index_inst->type_id(),
+                                            last_index_inst->result_id(),
+                                            element_inst->result_id());
+    new_value_id = addition->result_id();
+  } else {
+    // Indexing into structs must be constant, so bail out here.
+    return false;
```

Comment:
This is defensive, right?   A valid module would never get here, I think.?

Reply:
Not entirely. It covers the case where I have:
```
%gep = OpAccessChain %ptr <indices to a struct>
%ptr_ptr = OpPtrAccessChain %gep %non_const_element <indices>
```
The slide can't be converted properly.